0000-access-keys.md

```diff
@@ -0,0 +1,313 @@
+- Proposal Code Name: access_keys
+- Start Date: 2019-07-08
+- NEP PR: [nearprotocol/neps#0000](https://github.com/nearprotocol/neps/pull/0000)
+- Issue(s): [nearprotocol/nearcore#687](https://github.com/nearprotocol/nearcore/issues/687)
+
+# Summary
+[summary]: #summary
+
+Access keys provide limited access to an account.
+Each access key belongs to some account and identified by a unique (within the account) public key.
+One account may have large number of access keys.
+Access keys will replace original account-level public keys.
+Access keys allow to act on behalf of the account by restricting allowed transactions with the access key permissions.
+
+# Motivation
+[motivation]: #motivation
+
+Access keys give an ability to use dApps in a secure way without asking the user to sign every transaction in the wallet.
+By issuing the access key once for the application, the application can now act on behalf of the user in a restricted environment.
+This enables seamless experience for the user.
+
+Access keys also enable a few other use-cases that are discussed in details below.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Here are proposed changes for the AccessKey and Account structs.  
+
+```rust
+/// `account_id,public_key` is a key in the state
+struct AccessKey {
+  /// The nonce for this.
+  pub nonce: Nonce,  // u64 
+  
+  /// Defines permissions for the AccessKey 
+  pub permission: AccessKeyPermission,
+}
+
+/// Defines permissions for AccessKey 
+pub enum AccessKeyPermission {
+  /// Restricts AccessKey to only be used for function calls.
+  FunctionCall(FunctionCallPermission),
+
+  /// Gives full access to the account.
+  /// NOTE: It's used to replace account-level public keys.
+  FullAccess,
+}
+
+pub struct FunctionCallPermission {
+  /// The amount that can be spent for transaction fees by this access key from the account balance.
+  /// When used, both account balance and the allowance is decreased.
+  pub allowance: Balance,  // u128
+
+  /// The AccountID of the receiver of the transaction. The access key will restrict transactions to
+  /// only this receiver.
+  pub receiver_id: AccountId,  // String
+  
+  /// If not None, the access key would be restricted to call only the given method name.
+  pub method_name: Option<String>,
+}
+
+/// NOTE: This change removes account-level nonce and public keys.
+/// Key is `account_id`
+struct Account {
```

Comment:
We need to spec out here how transactions from different access keys are going to be ordered with respect to each other. For example:
3 access keys (A,B,C) issue 3 transactions each:
A1, A2, A3; B1,B2,B3; C1, C2, C3;
All these transactions operate on the same state so they need to have an order. First transaction to execute is one of {A1,B1,C1} that has lowest hash, let's say it is B1. Second transaction to execute is one of {A1,B2,C1} with lowest hash, etc.

Reply:
Good point. It should be similar to how we order transactions with conflicting nonces. E.g. which transaction is taken first for a given `account_id` key. Now with the access key we should do the same but for the `account_id,public_key` keys.

Reply:
I am not sure I understand this:

>E.g. which transaction is taken first for a given account_id key. Now with the access key we should do the same but for the account_id,public_key keys.

Reply:
Is there a reason to ever accept a transaction with `new_nonce != previous_nonce + 1` ?

Reply:
When transaction is lost or delayed due to how transactions are propagated through gossip. Should be a very common thing.

Reply:
But this is exactly what we can avoid by having a strict requirement. Asking this question in discord for broader discussion

Reply:
Copying from discord channel:

Transactions are propagated asynchronously through p2p network. If someone issues two transactions with nonces X+1 and X+2 (where X is the current nonce of the account) but X+1 is getting delayed we want to allow X+2 to be accepted and processed and X+1 to be discarded later. (your suggested rule would disallow this)
Consider a more severe example: Let's say someone issues 10 TPS (they do diligent work incrementing nonce correctly) and we are producing 1 block per second. If every 1 second at least one transaction is getting delayed by p2p network by 1 second then we are going to have a 50% failure rate of all transactions issued by this person on average.

Reply:
Added it to unresolved question. The actual TX ordering is not a blocking question, since it should just derive it from the current TX ordering with key upgrade.